core/src/smartcontracts/wasm/codec/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+/// This method takes ownership of the given pointer.
+#[allow(clippy::expect_used, clippy::unwrap_in_result)]
+pub fn decode_with_length_prefix_from_memory<
+    C: wasmtime::AsContextMut,
+    T: DecodeAll + std::fmt::Debug,
+>(
+    memory: &wasmtime::Memory,
+    dealloc_fn: &wasmtime::TypedFunc<(WasmUsize, WasmUsize), ()>,
+    mut context: &mut C,
+    offset: WasmUsize,
+) -> Result<T, Trap> {
+    const U32_TO_USIZE_ERROR_MES: &str = "`u32` should always fit in `usize`";
+
+    let len_size_bytes: u32 = core::mem::size_of::<WasmUsize>()
+        .try_into()
+        .expect("`u32` size should always fit into `u32`");
+    let len = u32::from_le_bytes(
+        memory.data(&mut context)[offset as usize..(offset + len_size_bytes) as usize]
+            .try_into()
+            .expect("Prefix length size(bytes) incorrect"),
+    );
+
+    let bytes = &memory.data_mut(&mut context)[offset.try_into().expect(U32_TO_USIZE_ERROR_MES)
+        ..(offset + len).try_into().expect(U32_TO_USIZE_ERROR_MES)];
+
+    let obj =
+        T::decode_all(&mut &bytes[len_size_bytes.try_into().expect(U32_TO_USIZE_ERROR_MES)..])
+            .map_err(|err| Trap::new(format!("Failed to decode object: {err}")))?;
+
+    dealloc_fn
+        .call(&mut context, (offset, len))
+        .map_err(|err| Trap::new(format!("Failed to call dealloc fn: {err}")))?; // TODO
+    Ok(obj)
+}
+
+/// Encode `obj` to the given `memory` with the given `alloc_fn` and `context`.
+///
+/// Returns the offset of the encoded object.
+pub fn encode_into_memory<T: Encode>(
+    obj: &T,
+    memory: &wasmtime::Memory,
+    alloc_fn: &wasmtime::TypedFunc<WasmUsize, WasmUsize>,
+    mut context: impl wasmtime::AsContextMut,
+) -> Result<WasmUsize, Trap> {
+    let bytes = encode_with_length_prefix(obj);
+
+    let len = bytes
+        .len()
+        .try_into()
+        .expect("Checked in `encode_with_length_prefix`");
+
+    let offset = alloc_fn.call(&mut context, len)?;
+    let offset_usize = offset
+        .try_into()
+        .expect("`u32` should always fit in `usize`");
+
+    memory
+        .write(&mut context, offset_usize, &bytes)
+        .map_err(|err| Trap::new(err.to_string()))?;
+
+    Ok(offset)
+}
+
+/// Encode the given object but also add it's length in front of it. This can be considered
+/// a custom encoding format.
+///
+/// Usually, to retrieve the encoded object both pointer and the length of the allocation
+/// are provided. However, due to the lack of support for multivalue return values in stable
+/// `WebAssembly` it's not possible to return two values from a wasm function without some
+/// shenanignas. In those cases, only one value is sent which is pointer to the allocation
+/// with the first element being the length of the encoded object following it.
+pub fn encode_with_length_prefix<T: Encode>(obj: &T) -> Vec<u8> {
+    // Compile-time size check
+    #[allow(clippy::let_unit_value)]
+    let _ = <SizeChecker<T> as DoesSizeFitInWasmUsize>::RESULT;
```

Comment:
this is cool

Reply:
is this const expression guaranteed to be evaluated. Does it matter that the result is bound to a placeholder?

Reply:
The author of the article I mentioned doesn't do that but ok

Reply:
I get this error:

```
error[E0401]: can't use generic parameters from outer function
   --> core/src/smartcontracts/wasm/codec/src/lib.rs:123:38
    |
120 | pub fn encode_with_length_prefix<T: Encode>(obj: &T) -> Vec<u8> {
    |                                  - type parameter from outer function
...
123 |     const RESULT: () = SizeChecker::<T>::RESULT;
    |                                      ^ use of generic parameter from outer function
```

Very strange. Let's keep it like it's already done

Reply:
The best description I've found:

> The error message is not the best, but items like fns and consts inside a function definition behave as if they were actually declared outside the function. The only difference is that their visibility is restricted to inside the containing function.

Looks like it possible only with a nightly feature.

[source](https://users.rust-lang.org/t/error-e0401-cant-use-generic-parameters-from-outer-function/84512)